### PR TITLE
[SPARK-11980] [SQL] Fix json_tuple and add test cases for SPARK-10621

### PR DESCRIPTION
**Pull request by gatorsmile**
Added Python test cases for the function `isnan`, `isnull`, `nanvl` and `json_tuple`. 

Fixed a bug in the function `json_tuple`  

@rxin , could you help me review my changes? Please let me know anything is missing. 

Thank you! Have a good Thanksgiving day!

### CODE REVIEW

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -305,6 +297,10 @@ def input_file_name():
 @since(1.6)
 def isnan(col):
     """An expression that returns true iff the column is NaN.
+
+    >>> df = sqlContext.createDataFrame([(1.0, float('nan')), (float('nan'), 2.0)], ("a", "b"))
+    >>> df.select(isnan("a").alias("r")).collect()
```
Comment (rxin) [inline python/pyspark/sql/functions.py]: can you test both string names and columns? e.g.
```python
df.select(isnan("a").alias("r1"), isnan(df.a).alias("r2")).collect()
```

Comment (rxin) [inline python/pyspark/sql/functions.py]: and do the same thing for the rest of the functions

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -1475,15 +1487,29 @@ def get_json_object(col, path):
...
-    sc = SparkContext._active_spark_context
-    jc = sc._jvm.functions.json_tuple(_to_java_column(col), fields)
+    >>> data = [("1", '''{"f1": "value1", "f2": "value2", "f3": 3, "f5": 5.23}'''), \
+          ("2", '''{"f1": "value12", "f3": "value3", "f2": 2, "f4": 4.01}'''), \
+          ("3", '''{"f1": "value13", "f4": "value44", "f3": "value33", "f2": 2, "f5": 5.01}'''), \
+          ("4", None), \
+          ("5", '''{"f1": "", "f5": null}'''), \
+          ("6", '''[invalid JSON string]''')]
+    >>> df = sqlContext.createDataFrame(data, ("key", "jstring"))
+    >>> df.select(df.key, json_tuple(df.jstring, 'f1', 'f2', 'f3', 'f4', 'f5')).collect()
```
Comment (rxin) [inline python/pyspark/sql/functions.py]: having a case with only two fields would be better. this one is too complicated as an example right now.

Comment (rxin) [inline python/pyspark/sql/functions.py]: while you at it, can you also add test for get_json_object?

Comment (gatorsmile) [inline python/pyspark/sql/functions.py]: sure, will do.

**Comment by rxin** [comment]
cc @davies for a final look.

The changes LGTM.

**Comment by gatorsmile** [comment]
In the next few days, I will look at the implementation of `get_json_object`. I suspect the original implementation has an issue regarding null values. 

If you compare the results of `get_json_object` and `json_tuple`, one attribute is different. I will try to fix it in a separate PR. This should not be related to the Python interface, I think. 

Thank you!

**Comment by gatorsmile** [comment]
Just did a quick check. I can confirm this is not caused by Python. I reproduced it using the scala API.

**Comment by gatorsmile** [comment]
Narrowed down to the following code in jsonExpressions.scala:
```scala
          val output = new ByteArrayOutputStream()
          val matched = Utils.tryWithResource(
            jsonFactory.createGenerator(output, JsonEncoding.UTF8)) { generator =>
            parser.nextToken()
            evaluatePath(parser, generator, RawStyle, parsed.get)
          }
```

So far, our parser returns the same results `output` for the following two cases. Both results are `"null"`:

```scala
    val tuple: Seq[(String, String)] = ("5", """{"f1": null}""") :: Nil
    val df: DataFrame = tuple.toDF("key", "jstring")
    val res = df.select(functions.get_json_object($"jstring", "$.f1")).collect()
```
```scala
    val tuple2: Seq[(String, String)] = ("5", """{"f1": "null"}""") :: Nil
    val df2: DataFrame = tuple2.toDF("key", "jstring")
    val res3 = df2.select(functions.get_json_object($"jstring", "$.f1")).collect()
```

**Comment by gatorsmile** [comment]
Found a discussion about this issue:

http://www.scriptscoop.net/t/1a9222820510/java-how-to-tell-jackson-to-deserialize-null-string-to-null-literal.html

Please let me know what I should do next. Thanks! @rxin @davies @marmbrus @cloud-fan

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -1468,22 +1481,47 @@ def get_json_object(col, path):
...
     :param fields: list of fields to extract
 
+    >>> data = [("1", '''{"f1": "value1", "f2": "value2"}'''), \
+          ("2", '''{"f1": "value12"}'''), \
+          ("3", '''{"f2": 2}'''), \
+          ("4", None), \
+          ("5", '''{"f1": null}'''), \
+          ("6", '''[invalid JSON string]''')]
+    >>> df = sqlContext.createDataFrame(data, ("key", "jstring"))
+    >>> df.select(df.key, json_tuple(df.jstring, 'f1', 'f2')).collect()
```
Comment (davies) [inline python/pyspark/sql/functions.py]: nit: I think one simple case should be enough for Python tests, other corner cases should be tested in Scala.

The Python doc tests will be part of API doc, so it's better to be read friendly.

Comment (gatorsmile) [inline python/pyspark/sql/functions.py]: Sure, will do. I will move the test case of `get_json_object` to the scala test file. Will simplify the existing test cases of `get_json_object` and `json_tuple`. Thanks!

**Comment by rxin** [comment]
I'd just simply the test case as Davies suggested, and then merge this in. In parallel you can work on a patch to fix whatever bugs you find.

**Comment by rxin** [comment]
Thanks - I'm going to merge this.

**Comment by gatorsmile** [comment]
Thank you for your help! @rxin @davies 

Just let me know if you need me to do any JIRA. Have a good holiday!

**Pull request closed without merging**